File: flink-sample-scala/src/main/scala/udf/TemperatureAverager.scala
package udf

import org.apache.flink.streaming.api.scala.function.WindowFunction
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector
import util.SensorReading

class TemperatureAverager extends WindowFunction[SensorReading, SensorReading, String, TimeWindow] {
  override def apply(key: String, window: TimeWindow, input: Iterable[SensorReading], out: Collector[SensorReading]): Unit = {
    val (cnt, sum) = input.foldLeft((0, 0.0))((c, r) => (c._1 + 1, c._2 + r.temperature))
    val avgTemp = sum / cnt
    out.collect(SensorReading(key, window.getEnd, avgTemp))
  }
}

File: flink-datastream-1.7.1/src/main/java/window/TemperatureAverager.java
package window;

import org.apache.flink.streaming.api.functions.windowing.WindowFunction;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;
import util.SensorReading;

public class TemperatureAverager implements WindowFunction<SensorReading, SensorReading, String, TimeWindow> {
    @Override
    public void apply(String s, TimeWindow timeWindow, Iterable<SensorReading> iterable, Collector<SensorReading> collector) throws Exception {
        int cnt = 0;
        double sum = 0.0;
        for (SensorReading r : iterable) {
            cnt++;
            sum += r.temperature;
        }
        double avgTemp = sum / cnt;
        collector.collect(new SensorReading(s, timeWindow.getEnd(), avgTemp));
    }
}
